js-packages/@prestojs/rest/src/useAsyncValue.ts

```diff
@@ -0,0 +1,249 @@
+): UseAsyncValueReturn<T>;
+/**
+ * Resolve values from an array of ids using an async function.
+ *
+ * For the specified array of `ids` the `resolve` function will be called
+ * and should return an array of the same size with each entry being the resolved
+ * value for the corresponding id.
+ *
+ * If `existingValues` is provided this will be consulted first before calling
+ * `resolve`. This is useful when you have a list of values that may or
+ * may not have the data you care about. If it's there then the values will
+ * be returned immediately. Note that if any of the ids are missing from
+ * `existingValues` then it will be ignored and a call to `resolve` will be made
+ * requesting values for all `ids`.
+ */
+export default function useAsyncValue<T, U extends Id>(
+    props: UseAsyncValuePropsMulti<T, U>
+): UseAsyncValueReturn<T[]>;
+export default function useAsyncValue<T, U extends Id>(
+    props: CommonProps<T, U> & {
+        id?: U | null;
+        ids?: U[] | null;
+        resolve: (idOrIds: U | U[]) => Promise<T[] | T>;
+    }
+): UseAsyncValueReturn<T | T[]> {
+    const { id, ids, getId, resolve, trigger } = props;
+    if (id && ids) {
+        throw new Error("Only one of 'id' and 'ids' should be provided");
+    }
+    const idRef = useRef(id || ids);
+    const resolveRef = useRef(resolve);
+    const cache = useRef(new Map<string, T>());
+    resolveRef.current = resolve;
+    let valueInChoices;
+    const existingValues = props.existingValues || [];
+    if (id || ids) {
+        if (ids) {
+            const existingIds = existingValues.reduce((acc, item) => {
+                acc[hashId(identifiableGetId(item, getId))] = item;
+                return acc;
+            }, {});
+            valueInChoices = [];
+            for (const id of ids) {
+                const hashedId = hashId(id);
+                const item = existingIds[hashedId];
+                if (item) {
+                    valueInChoices.push(item);
+                } else {
+                    const cachedItem = cache.current.get(hashedId);
+                    if (cachedItem) {
+                        valueInChoices.push(cachedItem);
+                    }
+                }
+            }
+            if (valueInChoices.length !== ids.length) {
+                valueInChoices = null;
+            }
+        } else {
+            for (const item of existingValues) {
+                if (identifiableGetId(item, getId) === id) {
+                    valueInChoices = item;
+                    break;
+                }
+            }
+            if (!valueInChoices && id) {
+                const hashedId = hashId(id);
+                const cachedItem = cache.current.get(hashedId);
+                if (cachedItem) {
+                    valueInChoices = cachedItem;
+                }
+            }
+        }
+    }
+    const isValueMissing = (id || ids) && !valueInChoices;
+    const execute = useCallback(id => resolveRef.current(id), []);
```

Comment:
Nothing invalidates this because it has empty deps. It reads the ref at execution time so that `execute` never changes

`cache` is a ref - it will be freed when the component unmounts